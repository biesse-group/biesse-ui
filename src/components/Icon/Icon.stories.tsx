import { type Meta, type StoryObj } from "@storybook/react";

import { Icon, type IconProps } from "./Icon";

const meta: Meta<typeof Icon> = {
  title: "Icons/Icon",
  component: Icon,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Icon>;

const defaultArgs: IconProps = {
  name: "careers",
};

export const Large: Story = {
  args: {
    ...defaultArgs,
    size: "lg",
  },
};

export const Medium: Story = {
  args: {
    ...defaultArgs,
    size: "md",
  },
};

export const Small: Story = {
  args: {
    ...defaultArgs,
    size: "sm",
  },
};

export const ExtraSmall: Story = {
  args: {
    ...defaultArgs,
    size: "xs",
  },
};
