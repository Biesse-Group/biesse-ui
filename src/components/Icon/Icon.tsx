import { FC } from "react";
import styled from "styled-components";

import iconsMap from "./icons-map";

export type IconName = keyof typeof iconsMap;

export interface IconProps {
  /**
   * Optional component class name
   */
  className?: string;
  /**
   * Id of the icon you want to use
   */
  name: IconName;
  /**
   * Which from the default sizes of the icon you want to use
   */
  size?: "xs" | "sm" | "md" | "lg" | string;
  /**
   * Color fo the icon.
   * if not specified becomes inherit.
   */
  color?: string;
  testId?: string;
  className?: string;
}

const getIconSize = (size?: IconProps["size"]) => {
  switch (size) {
    case "xs":
      return "20px";
    case "sm":
      return "30px";
    case "lg":
      return "80px";
    case "md":
      return "50px";
    default:
      return size;
  }
};

const IconRoot = styled.div<Omit<IconProps, "name">>`
  color: ${(props) => props.color || "inherit"};

  > svg {
    height: ${(props) => getIconSize(props.size)};
    color: ${(props) => props.color || "inherit"};
  }
`;

export const Icon: FC<IconProps> = ({ name, size, color, className, testId, ...props }) => {
  const IconComponent = iconsMap[name];
  return (
    <IconRoot size={size} color={color} data-testid={testId} className={className}>
      <IconComponent {...props} />
    </IconRoot>
  );
};
