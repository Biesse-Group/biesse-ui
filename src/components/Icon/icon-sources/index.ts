export * from "./ArrowLeft";
export * from "./ArrowRight";
export * from "./Call";
export * from "./Careers";
export * from "./ChevronRight";
export * from "./Close";
export * from "./Contacts";
export * from "./Country";
export * from "./Customer";
export * from "./Download";
export * from "./Email";
export * from "./Hamburger";
export * from "./Innovation";
export * from "./Location";
export * from "./MaterialComposite";
export * from "./MaterialGlass";
export * from "./MaterialMetal";
export * from "./MaterialStone";
export * from "./MaterialWood";
export * from "./Picture";
export * from "./Play";
export * from "./FastRepair";
export * from "./RoundAngle";
export * from "./Search";
export * from "./SpareParts";
export * from "./Sustainability";
export * from "./VideoPill";
export * from "./Linkedin";
