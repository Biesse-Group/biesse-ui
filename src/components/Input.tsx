import { FC, useState } from "react";
import styled, { css } from "styled-components";

import { inputStyles } from "../styles/input-styles";
import { Button } from "./Button";

export type InputProps = {
  /**
   * Input placeholder shown when has no value
   */
  placeholder?: string;
  /**
   * Input current value
   */
  value?: string;
  /**
   * Input value change callback
   */
  onChange?: () => void;
  /**
   * Whether to show a dark or light shadow on focus/active state (default is `dark`)
   */
  shadow?: "dark" | "light";
  /**
   * Shows a submit button with the provided label
   */
  submitLabel?: string;
  /**
   * Called when submit button has been clicked
   */
  onSubmit?: (value: string) => void;
};

const StyledInput = styled.input<Omit<InputProps, "onSubmit">>`
  ${(props) => inputStyles(props.shadow)}

  &:active, &:focus {
    box-shadow: none;
  }

  ${(props) =>
    props.submitLabel &&
    css`
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    `}
`;

const InputContainer = styled.div<Pick<InputProps, "shadow"> & { hasFocus: boolean }>`
  position: relative;
  width: 100%;
  border-radius: ${(props) => props.theme.input.borderRadius};
  background-color: ${(props) => props.theme.color.white};
  display: flex;
  align-items: center;
  padding-right: 4px;
  transition: box-shadow 0.2s ease-out;

  ${(props) =>
    props.hasFocus &&
    css`
      box-shadow: 0 0 10px
        ${props.shadow === "dark"
          ? props.theme.input.boxShadow.dark
          : props.theme.input.boxShadow.light};
    `}
`;

const StyledButton = styled(Button)`
  height: 40px;
  padding-left: 15px;
  padding-right: 15px;
`;

export const Input: FC<InputProps> = ({ submitLabel, onSubmit, ...props }) => {
  const [focus, setFocus] = useState(false);

  return (
    <InputContainer shadow={props.shadow} hasFocus={focus}>
      <StyledInput {...props} onFocus={() => setFocus(true)} onBlur={() => setFocus(false)} />
      {submitLabel && (
        <StyledButton variant="primary" size="small">
          {submitLabel}
        </StyledButton>
      )}
    </InputContainer>
  );
};
