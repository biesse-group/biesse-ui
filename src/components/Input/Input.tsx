import { type FC, useState } from "react";
import styled, { css } from "styled-components";

import type { BaseProps } from "~components/baseProps";
import { inputStyles } from "~styles/input-styles";

export interface InputDecorationProps {
  value: string;
  focus: boolean;
}

export interface InputProps extends BaseProps {
  /**
   * Input placeholder shown when has no value
   */
  placeholder?: string;
  /**
   * Input current value
   */
  defaultValue?: string;
  /**
   * Input value change callback
   */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * Whether to show a dark or light shadow on focus/active state (default is `dark`)
   */
  shadow?: "dark" | "light";
  /**
   * Whether the input should have a border
   */
  border?: boolean;
  /**
   * Input Test ID
   */
  testId?: string;
  /**
   * Input type
   */
  type: React.InputHTMLAttributes<HTMLInputElement>["type"];
  /**
   * Optional decoration elements
   */
  startDecoration?: JSX.Element | ((props: InputDecorationProps) => JSX.Element);
  endDecoration?: JSX.Element | ((props: InputDecorationProps) => JSX.Element);
}

const StyledInput = styled.input<Pick<InputProps, "shadow">>`
  ${(props) => inputStyles(props.shadow)}

  &:active, &:focus {
    box-shadow: none;
  }

  border: none;
`;

const InputContainer = styled.div<Pick<InputProps, "shadow" | "border"> & { hasFocus: boolean }>`
  position: relative;
  width: 100%;
  border-radius: ${(props) => props.theme.input.borderRadius};
  background-color: ${(props) => props.theme.color.white};
  display: flex;
  align-items: center;
  padding-right: 4px;
  transition: box-shadow 0.2s ease-out;
  ${(props) =>
    props.border &&
    css`
      border: 1px solid ${props.theme.input.borderColor};
    `}

  ${(props) =>
    props.hasFocus &&
    css`
      box-shadow: 0 0 10px
        ${props.shadow === "dark"
          ? props.theme.input.boxShadow.dark
          : props.theme.input.boxShadow.light};
    `}
`;

export const Input: FC<InputProps> = ({
  testId,
  defaultValue = "",
  onChange,
  className,
  style,
  startDecoration,
  endDecoration,
  ...props
}) => {
  const [focus, setFocus] = useState(false);
  const [value, setValue] = useState(defaultValue);

  const { shadow, border } = props;

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    setValue(e.currentTarget.value);
    onChange?.(e);
  };

  const decorationsProps: InputDecorationProps = {
    value,
    focus,
  };

  return (
    <InputContainer hasFocus={focus} {...{ shadow, border, className, style }}>
      {typeof startDecoration === "function" ? startDecoration(decorationsProps) : startDecoration}
      <StyledInput
        {...props}
        onFocus={() => setFocus(true)}
        onBlur={() => setFocus(false)}
        onChange={handleChange}
        data-testid={testId}
      />
      {typeof endDecoration === "function" ? endDecoration(decorationsProps) : endDecoration}
    </InputContainer>
  );
};
