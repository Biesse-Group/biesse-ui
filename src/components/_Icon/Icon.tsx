import { FC } from "react";
import styled, { css, useTheme } from "styled-components";

import iconsMap from "./icons-map";

export interface IconProps {
  /**
   * Id of the icon you want to use
   */
  name: keyof typeof iconsMap;
  /**
   * Which from the default sizes of the icon you want to use
   */
  size?: "small" | "medium" | "large";
  /**
   * Custom size
   */
  height?: string;
  /**
   * Color of icon
   */
  color?: string;
}

const getIconSize = (size?: string) => {
  switch (size) {
    case "small":
      return css`
        height: 30px;
      `;
    case "large":
      return css`
        height: 80px;
      `;
    default:
      return css`
        height: 50px;
      `;
  }
};

const IconRoot = styled.div<Pick<IconProps, "size" | "height">>`
  > svg {
    height: 100%;
  }

  ${(props) =>
    props.height
      ? css`
          height: ${props.height};
        `
      : getIconSize(props.size)}
`;

export const Icon: FC<IconProps> = ({ name, size, height, ...props }) => {
  const IconComponent = iconsMap[name];
  const theme = useTheme();
  props.color = props.color || theme.color.primary;
  return (
    <IconRoot size={size} height={height}>
      <IconComponent {...props} />
    </IconRoot>
  );
};
