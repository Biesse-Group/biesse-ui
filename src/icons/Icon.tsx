import React, { FC } from "react";
import styled from "styled-components";

import getIconSize from "./get-icon-size";
import { IconProps } from "./IconProps";
import * as IconSources from "./icons-sources";

const IconRoot = styled.div<Omit<IconProps, "name">>`
  & svg {
    width: 100%;
  }

  ${(props) => getIconSize(props.size)}
`;

const renderIconImg = (name: string) => {
  switch (name) {
    case "search":
      return <IconSources.SearchIcon />;
    default:
      return <IconSources.PlaceHolderIcon />;
  }
};

export const Icon: FC<IconProps> = ({ name, ...props }) => {
  return <IconRoot {...props}>{renderIconImg(name)}</IconRoot>;
};
