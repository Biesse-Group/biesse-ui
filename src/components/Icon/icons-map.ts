import * as IconSources from "./icon-sources";

// Must be update each time a new icon is inserted in the library
const iconsMap = {
  arrow: IconSources.ArrowIcon,
  "arrow-right": IconSources.ArrowRightIcon,
  "arrow-left": IconSources.ArrowLeftIcon,
  careers: IconSources.CareersIcon,
  "material-wood": IconSources.CategoryWoodIcon,
  "material-metal": IconSources.CategoryMetalIcon,
  "material-composite": IconSources.CategoryCompositeIcon,
  "material-glass": IconSources.CategoryGlassIcon,
  "material-stone": IconSources.CategoryStoneIcon,
  "chevron-right": IconSources.ChevronRightIcon,
  contacts: IconSources.ContactsIcon,
  country: IconSources.CountryIcon,
  customer: IconSources.CustomerIcon,
  "fast-repair": IconSources.RepairFastIcon,
  hamburger: IconSources.HamburgerIcon,
  innovation: IconSources.InnovationIcon,
  location: IconSources.LocationIcon,
  "place-holder": IconSources.PlaceHolderIcon,
  search: IconSources.SearchIcon,
  sustainability: IconSources.SustainabilityIcon,
  "spare-parts": IconSources.SparePartsIcon,
};

export default iconsMap;
