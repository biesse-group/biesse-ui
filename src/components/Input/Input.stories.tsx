import { type Meta, type StoryObj } from "@storybook/react";
import { userEvent, within } from "@storybook/testing-library";
import styled from "styled-components";

import { Icon } from "~components/Icon";
import { BackgroundDecorator } from "~stories/decorators";

import { Input } from "./Input";

const meta: Meta<typeof Input> = {
  title: "Inputs/Input",
  component: Input,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
    docs: {},
  },
  argTypes: {
    onChange: {
      control: false,
    },
    error: {
      type: "boolean",
    },
  },
  decorators: [
    (Story, { args }) => (
      <BackgroundDecorator background={args.shadow === "light" ? "light" : "primary"}>
        <Story />
      </BackgroundDecorator>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof Input>;

export const DarkBackground: Story = {
  args: {
    placeholder: "Name",
    shadow: "dark",
  },
};

export const LightBackground: Story = {
  args: {
    ...DarkBackground.args,
    shadow: "light",
    border: true,
  },
};

export const WithStartDecoration: Story = {
  args: {
    ...DarkBackground.args,
    startDecoration: (
      <Icon color="primary" name="chevron-right" size="xs" style={{ marginLeft: 20 }} />
    ),
  },
};

export const WithEndDecoration: Story = {
  args: {
    ...DarkBackground.args,
    placeholder: "Cosa cerchi?",
    endDecoration: <Icon color="primary" name="search" size="24px" style={{ marginRight: 20 }} />,
  },
};

const StyledSubmitIcon = styled(Icon)`
  margin-right: 10px;
  :hover {
    cursor: pointer;
  }
`;

export const Filled: Story = {
  args: {
    ...DarkBackground.args,
    testId: "input",
    endDecoration: (
      <StyledSubmitIcon data-testId="submit" color="primary" name="chevron-right" size="24px" />
    ),
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await userEvent.type(canvas.getByTestId("input"), "John Doe", { delay: 100 });
    await userEvent.click(canvas.getByTestId("submit"));
  },
};

export const WithError: Story = {
  args: {
    ...DarkBackground.args,
    error: true,
    defaultValue: "An invalid value",
  },
};
