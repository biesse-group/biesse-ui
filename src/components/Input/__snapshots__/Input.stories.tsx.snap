// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputs/Input DarkBackground smoke-test 1`] = `
<div class="sc-gEvEer gcOOwi">
  <div style="max-width: 300px;">
    <div class="sc-fqkvVR hUAKJH"
         placeholder="Name"
    >
      <input placeholder="Name"
             class="sc-eqUAAy iLusnC"
      >
    </div>
  </div>
</div>
`;

exports[`Inputs/Input Filled play-test 1`] = `
<div class="sc-gEvEer gcOOwi">
  <div style="max-width: 300px;">
    <div class="sc-fqkvVR hUAKJH"
         placeholder="Name"
    >
      <input placeholder="Name"
             data-testid="input"
             class="sc-eqUAAy iLusnC"
      >
      <span color="primary"
            data-testid="submit"
            class="sc-aXZVg dUQDlI sc-dcJsrY iPuqvC"
      >
        <svg viewbox="0 0 50 50"
             xmlns="http://www.w3.org/2000/svg"
             xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs>
            <path d="M0 0h25.722v48H0z">
            </path>
          </defs>
          <g transform="translate(12 1)"
             fill="none"
             fill-rule="evenodd"
          >
            <path d="M1.768 48c-.26 0-.521-.043-.752-.16-1.134-.578-1.323-1.964-.543-2.807l21.408-20.659L.471 2.965A1.758 1.758 0 0 1 .522.518 1.756 1.756 0 0 1 1.77 0c.446 0 .872.167 1.2.47l22.21 22.56c.729.74.72 1.93-.02 2.657L2.97 47.53c-.324.302-.752.47-1.2.47h-.002Z"
                  fill="currentColor"
            >
            </path>
          </g>
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Inputs/Input LightBackground smoke-test 1`] = `
<div class="sc-gEvEer kJmFai">
  <div style="max-width: 300px;">
    <div class="sc-fqkvVR bJnDBG"
         placeholder="Name"
    >
      <input placeholder="Name"
             class="sc-eqUAAy jZRVaQ"
      >
    </div>
  </div>
</div>
`;

exports[`Inputs/Input WithEndDecoration smoke-test 1`] = `
<div class="sc-gEvEer gcOOwi">
  <div style="max-width: 300px;">
    <div class="sc-fqkvVR hUAKJH"
         placeholder="Cosa cerchi?"
    >
      <input placeholder="Cosa cerchi?"
             class="sc-eqUAAy iLusnC"
      >
      <span color="primary"
            class="sc-aXZVg dUQDlI"
            style="margin-right: 20px;"
      >
        <svg viewbox="0 0 50 50"
             xmlns="http://www.w3.org/2000/svg"
             xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs>
            <path d="M0 0h47.956v48H0z">
            </path>
          </defs>
          <g fill="none"
             fill-rule="evenodd"
          >
            <g transform="translate(1 1)">
              <path d="M20.579 3.389A17.278 17.278 0 1 1 3.3 20.668 17.3 17.3 0 0 1 20.58 3.389Zm0 37.901a20.42 20.42 0 0 0 13.377-4.927l11.148 11.148a1.672 1.672 0 0 0 2.363-2.363L36.319 34a20.425 20.425 0 0 0 4.927-13.377 20.624 20.624 0 1 0-20.623 20.623l-.044.044Z"
                    fill="currentColor"
              >
              </path>
            </g>
            <path d="M17.39 34.101a1.174 1.174 0 0 0 0-2.348 5.103 5.103 0 0 1-5.089-5.088 1.174 1.174 0 0 0-2.348 0 7.436 7.436 0 0 0 7.437 7.436"
                  fill="currentColor"
            >
            </path>
          </g>
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Inputs/Input WithError smoke-test 1`] = `
<div class="sc-gEvEer gcOOwi">
  <div style="max-width: 300px;">
    <div class="sc-fqkvVR haQUBC"
         placeholder="Name"
    >
      <input placeholder="Name"
             class="sc-eqUAAy iLusnC"
             value="An invalid value"
      >
    </div>
  </div>
</div>
`;

exports[`Inputs/Input WithStartDecoration smoke-test 1`] = `
<div class="sc-gEvEer gcOOwi">
  <div style="max-width: 300px;">
    <div class="sc-fqkvVR hUAKJH"
         placeholder="Name"
    >
      <span color="primary"
            class="sc-aXZVg dJGlPM"
            style="margin-left: 20px;"
      >
        <svg viewbox="0 0 50 50"
             xmlns="http://www.w3.org/2000/svg"
             xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs>
            <path d="M0 0h25.722v48H0z">
            </path>
          </defs>
          <g transform="translate(12 1)"
             fill="none"
             fill-rule="evenodd"
          >
            <path d="M1.768 48c-.26 0-.521-.043-.752-.16-1.134-.578-1.323-1.964-.543-2.807l21.408-20.659L.471 2.965A1.758 1.758 0 0 1 .522.518 1.756 1.756 0 0 1 1.77 0c.446 0 .872.167 1.2.47l22.21 22.56c.729.74.72 1.93-.02 2.657L2.97 47.53c-.324.302-.752.47-1.2.47h-.002Z"
                  fill="currentColor"
            >
            </path>
          </g>
        </svg>
      </span>
      <input placeholder="Name"
             class="sc-eqUAAy iLusnC"
      >
    </div>
  </div>
</div>
`;
